### DISCUSSION TITLE
Documentation for creating new data provider or extending existing

### DISCUSSION THREAD

**Discussion by DanielVernall** [linq2db/linq2db#4555] [Q&A]
Hello,

I'm looking to extend an existing data provider, or create a custom one. I'm aiming to achieve something similar to Linq2db.Remote but using a different transport. It is difficult to understand how Linq2db.Remote works as I can't find much documentation for it.
How is Linq2db.Remote used as it doesn't seem to work in the same way as other Linq2db data providers do, and the documentation is mainly for those.

**Comment by sdanyliv** [comment]
Well, it is subject to change. What you exectly is trying to do? I thought to use [Remote.Linq](https://github.com/6bee/Remote.Linq) for such purpose and deprecate/rewrite `Linq2db.Remote`. `Remote.Linq` should work with existing codebase.

**Reply by DanielVernall** [reply]
Thanks for the link to Remote.Linq.

We have software with its databases running locally on small business LANs which are not publicly addressable. To create a public API for the data, we host an ASP.NET core OData API. The software on the small business LAN then opens a SignalR connection to the API server. 

When requests come to the API, they are translated to SQL and a request is sent via the SignalR connection to the small business software which executes it locally and returns the result. 

This works currently, but OData to SQL translation is limited. OData in ASP.Net core works with IQueryable out-of-the-box however, so I have created a DataConnection/DataAdapter which uses our SignalRDataContext internally. This works in ADO.NET, but I can't seem to get it to work with a linq2db data provider.

**Reply by sdanyliv** [reply]
Which databases are under the hood?

**Reply by DanielVernall** [reply]
PostgreSQL

**Reply by sdanyliv** [reply]
I'm thinking about public API. What do you want to expose and for which client applications? If OData is not suitable, what will be suitable?

**Reply by DanielVernall** [reply]
The client desktop application is our company's main product and the public API is provided to enable 3rd party integrations, so OData provides a flexible REST API. The API is already in production using the current system with a subset of OData queries available to use, so we're looking to expand the API to support more OData features by using the IQueryable system built-in. 
We were hoping to use linq2db to leverage its Linq to SQL transformers from its data providers, but intercept the queries before they are executed in the database software and instead send them to the remote clients.

**Reply by sdanyliv** [reply]
I see the following call:
```
Desktop Client -> OData 
    Remote.Linq client (ClientDataRepository) which creates `IQueryable` - `remoreQueryable`
        OData.ApplyToQueryable(remoreQueryable)
             Remote.Linq calls Remote.Linq Serice (by ToList for example)

(On the one of the machines)
Remote.Linq Service accepts client's request
    Creates linq2db IQueryable, deserializes request's Expression
        Here you can intercept IQueryable and apply additional logic
              Query Executed and result serilized back to the client by Remote.Linq
       
```

**Reply by DanielVernall** [reply]
That's what I had been considering, if we were to use Remote.Linq, but I was hoping to avoid having to include linq2db on the local software, instead performing all of the transformation directly on the API and just passing SQL commands over the SignalR connection. 
But this is a viable option, assuming I can write an adapter for Remote.Linq to use a SignalR connection.

**Reply by sdanyliv** [reply]
You will not use `linq2db` on the client side. Remote.Linq provides it's own QueryProvider.

**Reply by DanielVernall** [reply]
I'm not sure I follow. In case I wasn't clear, here is the setup:

- Our desktop business software runs on machines in small business LANs.
- This desktop software connects to a locally running PostgreSQL server in the LAN. There is no internet facing server to access this database directly and it is only available within the LAN.
- We host a public API on our servers to give access to the business software's data.
- The desktop software opens an SignalR connection to the API server and waits for requests from it.
- When requests come to the API server from 3rd party cloud apps/webhooks/integrations etc. these requests are transformed into SQL and sent to the correct desktop software over the open SignalR connection. The desktop software executes the SQL on its local database connection and returns the result to the API server.
- The API server then includes the result in its HTTP response.

**Reply by DanielVernall** [reply]
The desktop software is legacy software which does not include an IQueryable interface to its database, instead using oldschool DataHandlers. So, we currently just send SQL to it which is executed and the result is serialized in to a DataTable.

**Reply by DanielVernall** [reply]
If we were to use Remote.Linq, as you suggest, it seems like we would need something to work with Linq in the desktop software to convert the Expression into SQL queries. Which is why I said we would need to include linq2db in the desktop software.

**Reply by DanielVernall** [reply]
The desktop software opens the bi-directional connection to the API server because the desktop software is not publicly accessible. So, in reality the desktop software is the **_server_** of the data, and the API server just acts as a gateway to the internet, for software to access the desktop software's database.

**Reply by sdanyliv** [reply]
I see. So you need to execute queries generated by `linq2db` but using remote server which returns DataTable. Right?

**Reply by DanielVernall** [reply]
Yes, that's correct.

**Reply by sdanyliv** [reply]
1. Yo need to implement new `RemotePostgreSQLDataProvider` derived from `DataProviderBase`. Copy flags from `PostgreSQLDataProvider`, You can reuse `PostgreSQLSqlBuilder`, `PostgreSQLSqlOptimizer`, and probably `PostgreSQLMappingSchema` also.
2. override `CreateConnectionInternal`, `CreateSqlBuilder`, `GetSqlOptimizer`.
3. You need to implement your own `DbConnection`, `DbCommand`, `DataReader` - It is part which should call remote clients.

**Reply by DanielVernall** [reply]
Thanks @sdanyliv, I'm sorry it has taken me so long to reply, I was working on different issues. This is very helpful. I'll let you know how it goes once I start implementing it.
